beacon-chain/state/types.go

```diff
@@ -0,0 +1,66 @@
+package state
+
+import (
+	"sync"
+
+	"github.com/protolambda/zssz/merkle"
+	pbp2p "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/prysmaticlabs/prysm/shared/stateutil"
+)
+
+// BeaconState defines a struct containing utilities for the eth2 chain state, defining
+// getters and setters for its respective values and helpful functions such as HashTreeRoot().
+type BeaconState struct {
+	state        *pbp2p.BeaconState
+	lock         sync.RWMutex
+	merkleLayers [][][]byte
+}
+
+// Initialize --
+func Initialize(st *pbp2p.BeaconState) (*BeaconState, error) {
+	fieldRoots, err := stateutil.ComputeFieldRoots(st)
+	if err != nil {
+		return nil, err
+	}
+	layers := merkleize(fieldRoots)
+	return &BeaconState{
+		state:        st,
```

Comment:
should we be proto cloning here when we initialize? Otherwise something outside of this package can mutate the value at that original state pointer @nisdas